components/servo/generic/servo.go

```diff
@@ -0,0 +1,337 @@
+	}
+	return deps, nil
+}
+
+const model = "generic"
+
+func init() {
+	registry.RegisterComponent(servo.Subtype, model,
+		registry.Component{
+			Constructor: newGenericServo,
+		})
+	config.RegisterComponentAttributeMapConverter(servo.SubtypeName, model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var attr servoConfig
+			return config.TransformAttributeMapToStruct(&attr, attributes)
+		},
+		&servoConfig{})
+}
+
+type servoGeneric struct {
+	generic.Unimplemented
+	pin       board.GPIOPin
+	min       float64
+	max       float64
+	logger    golog.Logger
+	opMgr     operation.SingleOperationManager
+	frequency uint
+	minUs     uint
+	maxUs     uint
+	pwmRes    uint
+	currPct   float64
+}
+
+func newGenericServo(ctx context.Context, deps registry.Dependencies, cfg config.Component, logger golog.Logger) (interface{}, error) {
+	logger.Info("building servo!!")
+	attr, ok := cfg.ConvertedAttributes.(*servoConfig)
+	if !ok {
+		return nil, utils.NewUnexpectedTypeError(&servoConfig{}, cfg.ConvertedAttributes)
+	}
+	boardName := attr.Board
+	b, err := board.FromDependencies(deps, boardName)
+	if err != nil {
+		return nil, errors.Wrap(err, "board doesn't exist")
+	}
+	pin, err := b.GPIOPinByName(attr.Pin)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin")
+	}
+	frequency, err := pin.PWMFreq(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get servo pin pwm frequency")
+	}
+	if attr.Frequency != nil {
+		err = pin.SetPWMFreq(ctx, *attr.Frequency, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "error setting servo pin frequency")
+		}
+		frequency = *attr.Frequency
+	}
+	if frequency > 500 {
+		return nil, errors.Errorf("PWM frequencies should not be above 500Hz, have %d", frequency)
+	}
+	min := 0.0
+	max := 180.0
+	if attr.MinDeg != nil {
+		min = *attr.MinDeg
+	}
+	if attr.MaxDeg != nil {
+		max = *attr.MaxDeg
+	}
+	startPos := 0.0
+	if attr.StartPos != nil {
+		startPos = *attr.StartPos
+	}
+	minUs := minWidthUs
+	maxUs := maxWidthUs
+	if attr.MinWidthUS != nil {
+		minUs = *attr.MinWidthUS
+	}
+	if attr.MaxWidthUS != nil {
+		maxUs = *attr.MaxWidthUS
+	}
+	servo := &servoGeneric{
+		min:       min,
+		max:       max,
+		frequency: frequency,
+		pin:       pin,
+		logger:    logger,
+		minUs:     minUs,
+		maxUs:     maxUs,
+		currPct:   0,
+	}
+	if err := servo.Move(ctx, uint8(startPos), nil); err != nil {
+		return nil, errors.Wrap(err, "couldn't move servo to start position")
+	}
+	if servo.pwmRes == 0 {
+		if err := servo.findPWMResolution(ctx); err != nil {
+			return nil, errors.Wrap(err, "failed to guess the pwm resolution")
+		}
+		if err := servo.Move(ctx, uint8(startPos), nil); err != nil {
+			return nil, errors.Wrap(err, "couldn't move servo to start position")
+		}
+	}
+	return servo, nil
+}
+
+var _ = servo.LocalServo(&servoGeneric{})
+
+// Given minUs,maxUs,deg and frequency attempt to calculate the corresponding duty cylce pct.
```

Comment:
Mild preference for spaces after commas.

Reply:
👍